lsm/lsm.cpp

```diff
@@ -1,26 +1,125 @@
+  uint32_t seed = 1;
+  uint32_t key_hash;
+  MurmurHash3_x86_32(&k, sizeof k, seed, &key_hash);
   auto last_d = key_hash % 10;
   bset[last_d] = 1;
 }
-
 char BloomFilter::contains_key(int32_t k) {
-  auto key_hash = std::hash<int32_t>{}(k);
+  uint32_t seed = 1;
+  uint32_t key_hash;
+  MurmurHash3_x86_32(&k, sizeof k, seed, &key_hash);
   auto last_d = key_hash % 10;
   return bset[last_d];
 }
 
+int Level::size() {
+  return files.size();
+}
+void Level::add_file(File f) {
+  files.push_back(f);
+}
+std::string Level::next_file_path() {
+  return level_directory / std::to_string(files.size());
+}
+File Level::write_to_file(std::map<int32_t, maybe_value> m, std::string filename) {
+  std::ofstream ostrm(filename, std::ios::binary);
+  BloomFilter bloomFilter;
+  int min = std::numeric_limits<int>::max();
+  int max = std::numeric_limits<int>::lowest();
+  for (auto const&x : m) {
+    bloomFilter.insert_key(x.first);
+    ostrm.write(reinterpret_cast<const char*>(&x.first), sizeof x.first);
+    ostrm.write(reinterpret_cast<const char*>(&x.second.v), sizeof x.second.v);
+    ostrm.write(reinterpret_cast<const char*>(&x.second.is_deleted), sizeof x.second.is_deleted);
+    if (x.first > max) {max = x.first;}
+    if (x.first < min) {min = x.first;}
+  }
+  if (!ostrm) {
+    std::cerr << "could not write to file successfully" << "\n";
+    exit(5);
+  }
+  return File(filename, bloomFilter, min, max);
+}
+std::map<int32_t, maybe_value> Level::read_from_file(std::string filename) {
+  std::map<int32_t, maybe_value> new_m;
+  read_to_map(filename, new_m);
+  return new_m;
+}
+void Level::read_to_map(std::string filename, std::map<int32_t, maybe_value> &m) {
+  if (auto istrm = std::ifstream(filename, std::ios::binary)) {
+    int k;
+    while (istrm.read(reinterpret_cast<char *>(&k), sizeof k)) {
+      int v;
+      char d;
+      istrm.read(reinterpret_cast<char *>(&v), sizeof v);
+      istrm.read(reinterpret_cast<char *>(&d), sizeof d);
+      m.insert_or_assign(k, maybe_value(v,d));
+    }
+  } else {
+    std::cerr << "could not open " << filename << "\n";
+    exit(3);
+  }
+  std::cerr << "read file " << filename << " to map" << "\n";
+}
+File Level::merge(File older_f, File newer_f) {
+  std::map<int32_t, maybe_value> m;
+  read_to_map(older_f.filename, m);
+  read_to_map(newer_f.filename, m);
+  std::cerr << "merge files " << older_f.filename << "and" << newer_f.filename << " to new file " << "\n";
+  return write_to_file(m, next_file_path());
+}
+void Level::add(std::map<int32_t, maybe_value> m) {
+  File f = write_to_file(m, next_file_path());
+  add_file(f);
+}
+
+void LSM::add_to_level(std::map<int32_t, maybe_value> m, size_t l_index) {
+  Level& level = get_level(l_index);
+  if (level.size() + 1 == level.max_size) {
+    assert(level.max_size == 2);
+    File merged_f = level.merge(level.files.back(),
+                                          level.write_to_file(m, level.next_file_path()));
```

Comment:
align the two arguments
```suggestion
                                level.write_to_file(m, level.next_file_path()));
```